packages/js/components/src/tree-select-control/index.js

```diff
@@ -0,0 +1,497 @@
+			option.expanded
+				? nodesExpanded.filter( ( el ) => option.value !== el )
+				: [ ...nodesExpanded, option.value ]
+		);
+	};
+
+	/**
+	 * Handles a change on the Tree options. Could be a click on a parent option
+	 * or a child option
+	 *
+	 * @param {boolean}     checked Indicates if the item should be checked
+	 * @param {InnerOption} option  The option to change
+	 */
+	const handleOptionsChange = ( checked, option ) => {
+		if ( option.hasChildren ) {
+			handleParentChange( checked, option );
+		} else {
+			handleSingleChange( checked, option );
+		}
+
+		setInputControlValue( '' );
+		if ( ! nodesExpanded.includes( option.parent ) ) {
+			controlRef.current.focus();
+		}
+	};
+
+	/**
+	 * Handles a change of a child element.
+	 *
+	 * @param {boolean}     checked Indicates if the item should be checked
+	 * @param {InnerOption} option  The option to change
+	 */
+	const handleSingleChange = ( checked, option ) => {
+		const newValue = checked
+			? [ ...value, option.value ]
+			: value.filter( ( el ) => el !== option.value );
+
+		onChange( newValue );
+	};
+
+	/**
+	 * Handles a change of a Parent element.
+	 *
+	 * @param {boolean}     checked Indicates if the item should be checked
+	 * @param {InnerOption} option  The option to change
+	 */
+	const handleParentChange = ( checked, option ) => {
+		let newValue;
+		const changedValues = option.leaves
+			.filter( ( opt ) => opt.checked !== checked )
+			.map( ( opt ) => opt.value );
+
+		if ( checked ) {
+			if ( ! option.expanded ) {
+				handleToggleExpanded( option );
+			}
+			newValue = value.concat( changedValues );
+		} else {
+			newValue = value.filter( ( el ) => ! changedValues.includes( el ) );
+		}
+
+		onChange( newValue );
+	};
+
+	/**
+	 * Handles a change of a Tag element. We map them to Value format.
+	 *
+	 * @param {Array} tags List of current tags
+	 */
+	const handleTagsChange = ( tags ) => {
+		onChange( [ ...tags.map( ( el ) => el.id ) ] );
+	};
+
+	/**
+	 * Prepares and sets the search filter.
+	 * Filters of less than 3 characters are not considered, so we convert them to ''
+	 *
+	 * @param {Event} e Event returned by the On Change function in the Input control
+	 */
+	const handleOnInputChange = ( e ) => {
+		setInputControlValue( e.target.value );
+	};
+
+	return (
+		// eslint-disable-next-line jsx-a11y/no-static-element-interactions
+		<div
+			{ ...focusOutside }
+			onKeyDown={ onKeyDown }
+			className={ classnames(
+				'woocommerce-tree-select-control',
+				className
+			) }
+		>
+			{ !! label && (
+				<label
+					htmlFor={ `woocommerce-tree-select-control-${ instanceId }__control-input` }
+					className="woocommerce-tree-select-control__label"
+				>
+					{ label }
+				</label>
+			) }
+
+			<Control
+				ref={ controlRef }
+				disabled={ disabled }
+				tags={ getTags() }
+				isExpanded={ showTree }
+				onFocus={ () => {
+					setTreeVisible( true );
+				} }
+				onControlClick={ () => {
+					if ( disabled ) return;
```

Comment:
Looks like we already set the disabled value to `Control`, do we still need to handle the disabled state here?

Reply:
You're right. Refactor here https://github.com/woocommerce/woocommerce/pull/33384/commits/243068a233c335b5e0acf5b6b6e43d47593a518d